pyccel/stdlib/lists/lists.c

```diff
@@ -0,0 +1,207 @@
+        free((*list1)->elements);
+        *list1 = allocate_list(totalSize/tSizes[(*list1)->type], (*list1)->type, elements);
+        free(tmp);
+        free(elements);
+    }
+    else
+    {
+        elements = (char*)(*list1)->elements;
+        memcpy(&elements[(*list1)->size], list2->elements, list2->size);
+        (*list1)->size += tSizes[(*list1)->type];
+    }
+}
+
+void     clear(t_list* list)
+{
+    list->size = 0;
+}
+
+t_list*  copy(t_list* list)
+{
+    return (allocate_list(list->size / tSizes[list->type], list->type, list->elements));
+}
+
+size_t   count(t_list* list, void *item)
+{
+    size_t index = 0;
+    int count = 0;
+    char *elements = list->elements;
+    while (index < list->size)
+    {
+        if (list->type == lst_list)
+        {
+            t_list *tmp = (t_list*)&(elements[index]);
+            if (!(memcmp(item, tmp, sizeof(t_list))))
+                count += 1;
+        }
+        else if (!(memcmp(item, &(elements[index]), tSizes[list->type])))
+                count += 1;
+        index += tSizes[list->type];
+    }
+    return (count);
+}
+
+void     extend(t_list** list, void* item)
+{
+    size_t totalSize = (*list)->size + tSizes[(*list)->type];
+    char *elements = NULL;
+    if (totalSize > (*list)->capacity)
+    {
+        elements = malloc(totalSize);
+        memcpy(elements, (*list)->elements, (*list)->size);
+        memcpy(&elements[(*list)->size], item, 1);
+        *list = allocate_list(totalSize, (*list)->type, elements);
+        free(elements);
+    }
+    else
+    {
+        elements = (char*)(*list)->elements;
+        memcpy(&elements[(*list)->size], item, 1);
+    }
+}
+
+int   lst_index(t_list* list, void* item)
+{
+    size_t index = 0;
+    char *elements = (char*)(list->elements);
+    while (index < list->size)
+    {
+        if (list->type == lst_list)
+        {
+            t_list *tmp = (t_list*)(&elements[index]);
+            if (!(memcmp(item, tmp, sizeof(t_list))))
+                return index/tSizes[list->type];
+        }
+        else if (!(memcmp(item, &elements[index], tSizes[list->type])))
+                return index/tSizes[list->type];
+        index += tSizes[list->type];
+    }
+    return (-1);
+}
+
+void     insert(t_list** list, size_t index, void* item) // needs a redo
+{
+    size_t totalSize = (*list)->size + tSizes[(*list)->type];
+    index *= tSizes[(*list)->type];
+    if (index >= (*list)->size)
+        index = (*list)->size;
+    char *elements = NULL;
+    if (totalSize > (*list)->capacity)
+    {
+        elements = malloc(totalSize);
+        memcpy(elements, &(*list)->elements, index);
+        memcpy(&elements[index], item, tSizes[(*list)->type]);
+        memcpy(&elements[index + tSizes[(*list)->type]], (int8_t*)(*list)->elements + index * (*list)->type, (*list)->size - index);
+        *list = allocate_list(totalSize / (*list)->type, (*list)->type, elements);
```

Comment:
@azouiten It's not necessary to call `allocate_list` here as we only need to update `list->elements` to have new size. I think you can achieve this using `realloc` with the new size and shift the required bytes using `memcpy` or `memmove`.

Reply:
yes that works too, but I dont think realloc is the best choice here because it could fail if it doesnt find enough space to expand into.

Reply:
Well, `malloc` would fail too if it doesn't find enough space. The only reason why I thought of `realloc` is because it will copy the previous data for you, so you only need to call `memmove` on the bytes starting from `index`.